[kokseen1] iP

kokseen1 opened the pull request:
# Isla

> "The enemy of art is the absence of limitations." – Orson Welles ([source](https://quoteinvestigator.com/2014/05/24/art-limit/))

Isla frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/kokseen1/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Isla("./data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kokseen1 committed:
Implement Rename, Greet, Exit features

kokseen1 committed:
Remove all traces of Duke

kokseen1 committed:
Implement Echo feature

kokseen1 committed:
Implement Add and List features

kokseen1 committed:
Implement Mark and Unmark features

kokseen1 committed:
Implement ToDos, Events, Deadlines features

kokseen1 committed:
Implement Automated Text UI Testing

kokseen1 committed:
Implement exceptions to handle errors

kokseen1 committed:
Implement Delete

kokseen1 committed:
Update tests

kokseen1 committed:
Implement save and load features

kokseen1 committed:
Merge branch 'branch-Level-7'

kokseen1 committed:
Implement dates feature

kokseen1 committed:
Merge branch 'branch-Level-8'

kokseen1 committed:
Refactor to extract classes

kokseen1 committed:
Organize classes into packages

kokseen1 committed:
Merge pull request #1 from kokseen1/add-gradle-support

Add gradle support

kokseen1 committed:
Add Gradle support

kokseen1 committed:
Add JUnit tests

kokseen1 committed:
Add JavaDoc comments

kokseen1 committed:
Follow the Java coding standard

kokseen1 committed:
Implement find feature

kokseen1 committed:
Format code

kokseen1 committed:
Format code

kokseen1 committed:
Merge branch 'branch-A-CodingStandard'

kokseen1 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/isla/Parser.java
#	src/main/java/isla/Storage.java
#	src/main/java/isla/ui/Ui.java

kokseen1 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/isla/Parser.java

kokseen1 committed:
Fix wrong variable name

kokseen1 committed:
Use checkStyle to detect coding style violations

kokseen1 committed:
Add a GUI to the chatbot

kokseen1 committed:
Fix wrong chatbot name in variables

kokseen1 committed:
Clean up files used in old UI

kokseen1 committed:
Merge branch 'branch-Level-10'

samuelkwik commented on src/main/java/isla/Parser.java:
```diff
@@ -0,0 +1,198 @@
+            } catch (IndexOutOfBoundsException e) {
+                throw new IslaException("Target index is out of bounds.");
+            } catch (NumberFormatException e) {
+                throw new IslaException("Target index must be a number.");
```
I like the consistency of your error messages! 😄

samuelkwik commented on src/main/java/isla/task/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Enumerates and prints all the tasks in the list.
+     */
+    public String enumerate() {
```
Maybe a clearer method name can be used such as `getTasksAsString` which would make it clear that there is a string being returned?

samuelkwik commented on src/main/java/isla/task/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+    public Task deserialize(String serializedTask) throws IslaException {
+        String[] taskComponents = serializedTask.split("\\|");
+        String taskType = taskComponents[0];
+        boolean isDone = Boolean.parseBoolean(taskComponents[1]);
```
I like the use of `parseBoolean`!

samuelkwik commented on src/main/java/isla/task/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+     * Returns a TaskList with tasks matching the given keyword String.
+     */
+    public TaskList find(String keyword) {
+        return new TaskList((ArrayList<Task>) tasks.stream().filter(task -> task.description.toLowerCase()
```
I think the line length might be a bit long?

samuelkwik commented on src/main/java/isla/Parser.java:
```diff
@@ -0,0 +1,198 @@
+            throw new IslaException("Unknown command.");
+        }
+
+        storage.save(tasks.serialize());
```
Just some food for thought: do you need to save to storage for the find command? Additionally if there is a change in the taskList, but an error occurs while trying to save to storage, should the taskList remain updated?

samuelkwik reviewed (commented):
I found some useful methods that I definitely will be implementing in my own chatbot! Overall, very clearly structured code, other than some minor nits. LGTM 👍

jodielzy commented on src/main/java/isla/task/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+package isla.task;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
```
Would be neater if the imports are grouped in packages (space between java.time.* and java.util.* )?

jodielzy commented on src/main/java/isla/task/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Deletes the task at the given index.
+     */
+    public Task deleteTask(Integer taskId) {
```
I like the naming of the methods, they are clear and straightforward.

jodielzy commented on src/main/java/isla/ui/Ui.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Returns the chat delimiter message.
+     */
+    public String getLine() {
+        return "____________________________";
+    }
```
Would it be clearer to call this method "displayLine"?

jodielzy commented on src/main/java/isla/ui/Ui.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Returns the default farewell message of the chatbot.
+     */
+    public static String getFarewell() {
+        return "Bye. Hope to see you again.";
+    }
```
Would it be clearer to call this method "displayFarewell"?

jodielzy reviewed (commented):
Overall, I think that your code is well structured and nicely split into classes, the namings of classes and methods are also clear and straightforward.

kokseen1 committed:
Use Assertions

Certain assumptions in the code are crucial for correct execution
but are not explicitly checked, making debugging harder when
violations occur.

Failing to document these assumptions could lead to subtle bugs that
are difficult to trace, especially in cases where invalid states
propagate through the system without immediate failures.

Let's use Java's built-in assert feature to document these
assumptions at key points in the code. This ensures that violations
are detected early during development while having minimal runtime
impact in production (as assertions are typically disabled).

Assertions are preferred over explicit exception handling here
because they are intended for catching programming errors rather
than handling expected runtime conditions.

kokseen1 committed:
Improve Code Quality

Some parts of the codebase suffer from unnecessary complexity, poor
readability, and minor redundancies, making maintenance and
debugging more difficult.

Poorly structured code can lead to confusion and increase the
likelihood of introducing bugs when making future modifications.

Let's refactor the code to improve clarity, remove redundant logic,
and adhere to best practices discussed in the 'Code Quality' topics.
This includes:
* Replacing magic numbers with named constants for better
readability.
* Improving variable and method naming to enhance clarity and
maintainability.

These changes help create a more maintainable and robust codebase,
ensuring future modifications can be made with confidence.

kokseen1 committed:
Use Streams

kokseen1 committed:
Merge pull request #2 from kokseen1/branch-A-Assertions

Use Assertions

kokseen1 committed:
Merge branch 'master' into branch-A-CodeQuality

kokseen1 committed:
Merge branch 'master' into branch-A-Streams

kokseen1 committed:
Merge pull request #3 from kokseen1/branch-A-CodeQuality

Branch a code quality

kokseen1 committed:
Merge pull request #4 from kokseen1/branch-A-Streams

Branch a streams

kokseen1 committed:
Add help page

kokseen1 committed:
Merge branch 'branch-C-Help'

kokseen1 committed:
Add automatic resizing

kokseen1 committed:
Add window title

kokseen1 committed:
Improve error messages

kokseen1 committed:
Add product screenshot

kokseen1 committed:
Add greeting message

kokseen1 committed:
Remove help from greeting message

kokseen1 committed:
Fix comments to follow coding standard

kokseen1 committed:
Refactor parser and improve code quality

kokseen1 committed:
Fix missing newline

kokseen1 committed:
Use stream for enumeration

kokseen1 committed:
Refactor Storage and TaskList

kokseen1 committed:
Refactor TaskList

kokseen1 committed:
Fix out of bounds error

kokseen1 committed:
Improve output formatting

kokseen1 committed:
Improve comments according to conventions

kokseen1 committed:
Rename variable

kokseen1 committed:
Fix JavaDoc comments

kokseen1 committed:
Refactor save functionality

kokseen1 committed:
Fix TaskList tests

kokseen1 committed:
Fix test assertion

kokseen1 committed:
Add user guide

kokseen1 committed:
Fix user guide
